plugins/microchip/switchtec-nvme.c

```diff
@@ -399,3 +400,272 @@ static int switchtec_pax_list(int argc, char **argv, struct command *command,
+	if (ret)
+		return NOT_FOUND;
+
+	ret = switchtec_fab_gfms_db_dump_fabric_general(dev, &fg);
+	if (ret)
+		return NOT_FOUND;
+
+	for (i = 0; i < 16; i++) {
+		if (topo_info.route_port[i] == 0xff && fg.hdr.pax_idx != i)
+			continue;
+
+		r_type = fg.body.pax_idx[i].reachable_type;
+
+		if (fg.hdr.pax_idx == i ||
+		    r_type == SWITCHTEC_GFMS_DB_REACH_UC ||
+		    r_type == SWITCHTEC_GFMS_DB_REACH_BC) {
+			ret = switchtec_set_pax_id(dev, i);
+			if (ret)
+				continue;
+
+			ret = pax_check_ep_pdfid(dev, pdfid);
+			if (ret != NOT_FOUND)
+				break;
+
+		}
+	}
+
+	switchtec_set_pax_id(dev, SWITCHTEC_PAX_ID_LOCAL);
+	return ret;
+}
+
+#define PCIE_CAPS_OFFSET		0x34
+#define PCIE_CAPS_OFFSET_MASK		0xfc
+#define PCIE_CAP_ID			0x10
+#define PCIE_DEVICE_CTRL_OFFSET		0x08
+#define PCIE_DEVICE_RESET_FLAG		0x8000
+#define PCIE_NEXT_CAP_OFFSET		0x01
+
+int ask_if_sure(int always_yes)
+{
+	char buf[10];
+	char *ret;
+
+	if (always_yes)
+		return 0;
+
+	fprintf(stderr, "Do you want to continue? [y/N] ");
+	fflush(stderr);
+	ret = fgets(buf, sizeof(buf), stdin);
+
+	if (!ret)
+		goto abort;
+
+	if (strcmp(buf, "y\n") == 0 || strcmp(buf, "Y\n") == 0)
+		return 0;
+
+abort:
+	fprintf(stderr, "Abort.\n");
+	errno = EINTR;
+	return -errno;
+}
+
+static int switchtec_vf_reset(int argc, char **argv, struct command *command,
+			      struct plugin *plugin)
+{
+	int ret = 0;
+	uint8_t offset;
+	uint16_t pdfid;
+	uint32_t ns_id;
+	uint8_t cap_id = 0;
+	uint16_t flags;
+	char device_str[64];
+	char pdfid_str[64];
+	struct switchtec_dev *dev;
+	const char *desc = "Perform a Function Level Reset (FLR) on a Virtual Function";
+	const char *force = "The \"I know what I'm doing\" flag, skip confirmation before sending command";
+	struct config {
+		int force;
+	} cfg = {};
+
+	const struct argconfig_commandline_options opts[] = {
+		OPT_FLAG("force", 'f', &cfg.force, force),
+		{NULL}
+	};
+
+	ret = argconfig_parse(argc, argv, desc, opts);
+	if (ret < 0)
+		return ret;
+
+	if (optind >= argc) {
+		fprintf(stderr, "vf-reset: DEVICE is required for this command!\n");
+		return -1;
+	}
+
+	if (sscanf(argv[optind], "%2049[^@]@%s", pdfid_str, device_str) < 2) {
+		fprintf(stderr, "vf-reset: invalid device %s\n", argv[optind]);
+		fprintf(stderr, "\nNOTE: This command only supports devices in the form of 'PDFID@device',\n"
+				"where 'device' is your local device file or MOE service handle.\n"
+				"Example for local device: 0x3b01@/dev/switchtec0\n"
+				"Example for MOE access: 0x3b01@192.168.1.10:0\n");
+		fprintf(stderr, "\nFor RC device, use the reset command provided by your OS instead.\n"
+				"Example: echo 1 > /sys/bus/pci/devices/${bdf_vf}/reset\n");
+		return -1;
+	}
+
+	ret = sscanf(pdfid_str, "%hxn%d", &pdfid, &ns_id);
+	if (!ret) {
+		fprintf(stderr, "vf-reset: invalid device %s\n", argv[optind]);
+		return -1;
+	}
+
+	if (!cfg.force) {
+		fprintf(stderr, "WARNING: This will reset the Virtual Function with PDFID 0x%x\n\n",
```

Comment:
Maybe we can give the full device string like PDFID@/dev/switchte* in the error message. Think of the case in which a host connects to two PAX fabrics and both of them have the same PDFID.

Reply:
I still think adding device handle info in the warning message is helpful.

Reply:
Forgot to update this. Now it's updated.